cassandra/src/main/java/se/tre/freki/storage/cassandra/search/CassandraSearchPlugin.java

```diff
@@ -90,25 +96,120 @@ public void close() throws IOException {
   @Override
   public ListenableFuture<Void> deleteLabelMeta(final LabelId id,
                                                 final LabelType type) {
-    return Futures.immediateFuture(null);
+    return immediateFuture(null);
   }
 
   @Nonnull
   @Override
   public ListenableFuture<Void> indexAnnotation(final Annotation note) {
-    return Futures.immediateFuture(null);
+    return immediateFuture(null);
   }
 
   @Nonnull
   @Override
   public ListenableFuture<Void> deleteAnnotation(final Annotation note) {
-    return Futures.immediateFuture(null);
+    return immediateFuture(null);
   }
 
   @Nonnull
   @Override
   public ListenableFuture<Iterable<LabelMeta>> findLabels(final String query) {
-    return Futures.<Iterable<LabelMeta>>immediateFuture(ImmutableSet.<LabelMeta>of());
+    final NGramGenerator ngrams = new NGramGenerator(query);
+    final List<ResultSetFuture> resultSets = fetchLabelIds(ngrams);
+
+    return transform(
+        successfulAsList(resultSets),
+        new AsyncFunction<List<ResultSet>, Iterable<LabelMeta>>() {
+          @Override
+          public ListenableFuture<Iterable<LabelMeta>> apply(final List<ResultSet> resultSets) {
+            return transform(fetchLabelMetas(resultSets),
+                new Function<Iterable<LabelMeta>, Iterable<LabelMeta>>() {
+                  @Override
+                  public Iterable<LabelMeta> apply(final Iterable<LabelMeta> input) {
+                    return input;
+                  }
+                });
+          }
+        });
+  }
+
+  /**
+   * Takes a List of ResultSets and for each labelid in the resulset
+   * get the corresponding LabelMeta.
+   *
+   * @param resultSets A List of ResultSets
+   *
+   * @return A ListenableFuture with a list of LabelMetas.
+   */
+  protected ListenableFuture<? extends Iterable<LabelMeta>> fetchLabelMetas(
+      final List<ResultSet> resultSets) {
+
+    final List<ListenableFuture<LabelMeta>> metas = new ArrayList<>();
+
+    for (ResultSet resultSet : resultSets) {
+      for (Row row : resultSet) {
+        CassandraLabelId id = CassandraLabelId.fromLong(row.getLong("label_id"));
+        metas.add(cassandraStore.getMeta(id, id.type()));
+      }
+    }
+
+    return allAsList(metas);
+  }
+
+  /**
+   * Takes a NGramGenerator of a String and queries cassandra for labelids macthing
+   * each 3-gram in the NGramGenerator supplied.
```

Comment:
We should try to not talk about it being a 3-gram outside of the `NGramGenerator` since it is not exposed outside of that class. Can you change to "n-gram"?